src/histogram/axis.cpp

```diff
@@ -56,10 +55,17 @@ py::class_<A> register_axis_by_type(py::module& m, const char* name, const char*
     
     ;
     
+    // This could be a constexpr if, but no cost for being runtime (since pybind config runs once)
+    if(std::is_same<A, axis::category_str>::value) {
```

Comment:
More generically, you could check whether the bin method returns a reference and then use keep_alive.

Reply:
I am not sure why this should have caused you problems. According to docs, the default policy "return_value_policy::automatic" should have worked, although it is not efficient.
https://pybind11.readthedocs.io/en/stable/advanced/functions.html?highlight=keep_alive

Reply:
You need the keep_alive, otherwise when you run, for example, `repr(hist.axis(0).bin)`, the axis may be destroyed before the repr is called, causing it to produce `<bin [inf, inf]>`. It's  harder to reproduce in testing since pytest "remembers" the parts of the expression, but I did come up with a slightly more convoluted test that triggers the problem.

Reply:
> More generically, you could check whether the bin method returns a reference and then use keep_alive.

Is there a way to do that in C++14 that is not deprecated in C++17 and removed in C++20? In Boost perhaps?

Reply:
Nevermind, I got it without using `result_of`.